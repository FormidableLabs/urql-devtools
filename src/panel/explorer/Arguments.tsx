import React from "react";
import styled from "styled-components";
import { FieldNode } from "../context/explorer/ast";
import { SeeMoreIcon } from "./Icons";
import { KeyValue } from "./Value";

interface Props {
  args: FieldNode["args"];
  displayAll: boolean;
}

export function Arguments({ args, displayAll }: Props) {
  if (!args) {
    return null;
  }

  let content = [];
  const entries = Object.entries(args);

  for (const [key, val] of entries) {
    if (Array.isArray(val)) {
      content.push(
        <ArgWrapper key={key}>
          <ArgKey>{key}</ArgKey>
          <Symbol>{`:`}</Symbol>
          <IconContainer>
            {"["}
            <SeeMore />
            {"]"}
          </IconContainer>
          {content.length === entries.length - 1 ? "" : ","}
        </ArgWrapper>
      );
    } else if (val && typeof val === "object") {
      content.push(
        <ArgWrapper>
          <ArgKey>{key}</ArgKey>
          <Symbol>{`:`}</Symbol>
          <IconContainer>
            {"{"}
            <SeeMore />
            {"}"}
          </IconContainer>
          {content.length === entries.length - 1 ? "" : ","}
        </ArgWrapper>
      );
    } else {
      content.push(
        <ArgWrapper key={key}>
          <KeyValue keyName={key} value={val} expandValues={false} />
          {content.length === entries.length - 1 ? "" : ","}
        </ArgWrapper>
      );
    }
  }

  if (!displayAll && !(content.length <= 3)) {
    content = [
      ...content.slice(0, 3),
      <ExpandContainer>
        <SeeMore />
      </ExpandContainer>
    ];
  }

  return content ? <ArgsContainer>({content})</ArgsContainer> : null;
}

const ArgsContainer = styled.div`
  display: inline-flex;
  max-width: 400px;
  flex-wrap: wrap;
`;

const ArgKey = styled.span`
  color: ${p => p.theme.key};
`;

const ArgWrapper = styled.span`
  margin-right: 5px;

  &:last-of-type {
    margin-right: 0;
  }
`;

const IconContainer = styled.div`
  display: inline-flex;
  align-items: flex-end;
  color: ${p => p.theme.symbol};
  font-size: 14px;
  cursor: pointer;
`;

const ExpandContainer = styled(IconContainer)`
  margin: 5px 5px 0;
`;

const SeeMore = styled(SeeMoreIcon)`
  margin: 3px;
`;

const Symbol = styled.span`
  color: ${p => p.theme.symbol};
  margin-right: 5px;
`;
