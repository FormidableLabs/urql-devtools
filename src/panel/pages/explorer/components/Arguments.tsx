import React, { FC, useMemo, Fragment } from "react";
import styled from "styled-components";
import { ParsedFieldNode } from "../../../context/Explorer/ast";
import { InlineCodeHighlight } from "../../../components";

export const Arguments: FC<{
  args: ParsedFieldNode["args"];
}> = ({ args }) => {
  if (!args) {
    return null;
  }

  const entries = useMemo(() => Object.entries(args), [args]);

  return (
    <ArgumentText>
      (
      {entries.map(([key, value], index) => (
        <Fragment key={key}>
          {`${key}: `}
          <InlineCodeHighlight code={JSON.stringify(value) || "undefined" } language="json" />
          {index !== entries.length - 1 && ", "}
        </Fragment>
      ))}
      )
    </ArgumentText>
  );
};

const ArgumentText = styled.div`
  flex-basis: 0;
  flex-grow: 1;
  color: ${(p) => p.theme.grey["+4"]};
`;
