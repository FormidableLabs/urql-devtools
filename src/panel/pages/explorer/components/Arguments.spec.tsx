import React from "react";
import { shallow } from "enzyme";
import { Arguments } from "./Arguments"

const props = {
  args: {
    string: "abc",
    number: 7,
    object: {
      string: "def"
    },
    //JSON.stringify type definition is somewhat problematic.
    // JSON.stringify(null) //"null"
    // JSON.stringify(null) //undefined
    // But JSON.stringify return type is string.
    // Therefore, it will cause any TS null|undefined checks to actually fail at runtime.
    // THe Highlight component will crash if it ever gets a null|undefined code string.
    // Both cases have to be explicitly tested.
    nullable: null,
    undefinable: undefined
  }
}
describe("on mount", () => {
  it("matches snapshot", () => {
    const wrapper = shallow(<Arguments {...props}/>);
    expect(wrapper).toMatchSnapshot();
  })
})