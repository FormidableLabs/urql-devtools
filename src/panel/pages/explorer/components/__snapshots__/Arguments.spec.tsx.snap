// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on mount matches snapshot 1`] = `
<ArgumentText>
  (
  string: 
  <Memo(InlineCodeHighlightMemo)
    code="\\"abc\\""
    language="json"
  />
  , 
  number: 
  <Memo(InlineCodeHighlightMemo)
    code="7"
    language="json"
  />
  , 
  object: 
  <Memo(InlineCodeHighlightMemo)
    code="{\\"string\\":\\"def\\"}"
    language="json"
  />
  , 
  nullable: 
  <Memo(InlineCodeHighlightMemo)
    code="null"
    language="json"
  />
  , 
  undefinable: 
  <Memo(InlineCodeHighlightMemo)
    code="undefined"
    language="json"
  />
  )
</ArgumentText>
`;
